File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        ans = 0
        queue = deque([(root, 0)])
        
        while queue:
            size = len(queue)
            _, level_head_index = queue[0]
        
            for _ in range(size):
                node, col_index = queue.popleft()
                
                if node.left:
                    queue.append((node.left, 2 * col_index))
                
                if node.right:
                    queue.append((node.right, 2 * col_index + 1))
            
            ans = max(ans, col_index - level_head_index + 1)
        
        return ans

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
#define ull unsigned long long

class Solution {
    ull ans = 0;
    unordered_map<ull, ull> firstColIndexTable;
    
    void dfs(TreeNode* node, ull depth, ull colIndex){
        if (!node) return;
        if (firstColIndexTable.find(depth) == firstColIndexTable.end())
            firstColIndexTable[depth] = colIndex;
        ans = max(ans, colIndex - firstColIndexTable[depth] + 1);
        dfs(node -> left, depth + 1, 2 * colIndex);
        dfs(node -> right, depth + 1, 2 * colIndex + 1);
    }
    
public:
    int widthOfBinaryTree(TreeNode* root) {
        dfs(root, 0, 0);
        return ans;
    }
};